Design/histogram_type.md

```diff
@@ -0,0 +1,108 @@
+Despite this, there is currently no data type in Mantid that deals with histogram data.
+The purpose of this design proposal is to fix this.
+
+There are various motivations for this:
+
+- Reduce code duplication by providing arithmetic operations such as addition of histograms.
+- Hide handling of errors (uncertainties).
+- Hide discrimination between histogram and distribution data.
+- Decrease risk for certain types of bugs by using the C++ type system.
+
+All this would remove duplicated code from the client and put it into a central place where it can be unit-tested.
+
+
+## Current situation
+
+The current way of dealing with histograms has various shortcomings.
+Basically we have a set of three `std::vector<double>`, for `X`, `Y`, and `E`.
+There is a single `typedef` `MantidVec` used for all three of them (and other data), and this is probably adding only more confusion.
+
+The problems include:
+
+- No type safety:
+  - The bug in `ConvertUnits` creped in because we are currently using the same instance of a single type for both histogram data and distribution data. The actual meaning of a vector of `Y` values is basically kept as a mental flag in the brain of a Mantid developer.
+  - We are using a single type for `X`, `Y`, and `E`. It is thus easy to accidentally mix them up.
+  - The array `E` of error values is used for both the variance and the standard deviation. The state is again carried as a flag in the brain of a developer. We even have functions such as `rebin` that return an array of `E` values that may mean either variance or standard deviation depending on a flag to the function!
+- Code duplication: Re-implementing the same thing 10 times is wasteful and error prone.
+  - Operations like adding histograms are frequent.
+  - Propagation of errors.
+  - Conversions between variance and standard deviation.
+- Unit testing: Implementation of operations on histogram are currently mixed into algorithm code. Unit tests thus need to test two things at the same time.
+  - Many units tests of algorithms do not fully test the range of errors that can happen when, e.g., adding two spectra. Doing so would be very hard in the current de-centralized implementation,
+  - Many algorithms need to modify the errors. In some algorithms the unit tests cover the errors, in others they do not.
+
+
+## Design goals
+
+- Provide simple arithmetic operations for histograms.
+  - Arithmetic operations automatically deal with propagation of errors.
+- Hide differences between histogram and distribution data.
+  - Internal storage type could be either, does not need to be known at the interface level?
+  - Arithmetic operations should do the right thing for any data type automatically.
+- Hide error handling.
+  - For simple operations with histograms algorithm authors should not need to know how to propagate errors.
+- Enable unit testing of key concepts in Mantid.
+  - Arithmetic operations with histograms.
+  - Propagation of errors.
+
```

Comment:
Speed of both reads and writes should be preserved. Consider benchmarking the existing implementation in your roll-out steps.